backend/jack_backend.cpp

```diff
@@ -253,40 +328,74 @@ int JackMix::process( jack_nframes_t nframes, void* arg ) {
-                float max {0};
-		for ( jack_nframes_t n=0; n<nframes; n++ ) {
-                        tmp[ n ] *= volume;
-                        max = qMax(max, static_cast<float>(tmp[n]));
-                }
-                backend->newInputLevel(key, max);
+		QString key {in_it.key()};
+		jack_default_audio_sample_t* tmp = ins[ key ];
+		float volume = backend->getInVolume( key );
+		float volume_new = backend->getInVolumeNew( key );
+		float max {0};
+		if ( qFabs( ( volume_new - volume ) / volume ) > threshold) {
+			for ( jack_nframes_t n=0; n<nframes; n++ ) {
+				tmp[ n ] *=	volume + n * ( volume_new - volume ) / nframes;
+				max = qMax( max, static_cast<float>( tmp[ n ] ) );
+			}
+		} else {
+			for ( jack_nframes_t n=0; n<nframes; n++) {
+				tmp[ n ] *= volume_new;
+				max = qMax( max, static_cast<float>( tmp[ n ] ) );
+			}
+		}
+		backend->newInputLevel( key, max );
+		if ( volume_new != volume ) {
+			backend->setInVolume( key, volume_new );
+		}
 	}
+		 
 	/// The actual mixing.
 	for ( in_it = backend->in_ports.begin(); in_it != backend->in_ports.end(); ++in_it ) {
 		jack_default_audio_sample_t* tmpin = ins[ in_it.key() ];
 		for ( out_it = backend->out_ports.begin(); out_it != backend->out_ports.end(); ++out_it ) {
 			jack_default_audio_sample_t* tmpout = outs[ out_it.key() ];
-			float tmpvolume = backend->getVolume( in_it.key(), out_it.key() );
-			for ( jack_nframes_t n=0; n<nframes; n++ ) {
-				tmpout[ n ] += tmpin[ n ] * tmpvolume;
+			float volume = backend->getVolume( in_it.key(), out_it.key() );
+			float volume_new = backend->getVolumeNew( in_it.key(), out_it.key() );
+			if ( qFabs( ( volume_new - volume ) / volume ) > threshold ) {
+				for ( jack_nframes_t n=0; n<nframes; ++n ) {
+					tmpout[ n ] += tmpin[ n ] * ( volume + n * (volume_new - volume) / nframes );
+				}
+			} else {
+				for ( jack_nframes_t n=0; n<nframes; n++ ) {
+					tmpout[ n ] += tmpin[ n ] * volume_new;
+				}
+			}
+			if ( volume != volume_new) {
+				backend->setVolume( in_it.key(), out_it.key(), volume_new);
 			}
 		}
 	}
 	/// Adjust outlevels.
 	for ( out_it = backend->out_ports.begin(); out_it != backend->out_ports.end(); ++out_it ) {
-                QString key {out_it.key()};
+		QString key {out_it.key()};
 		jack_default_audio_sample_t* tmp = outs[ key ];
 		float volume = backend->getOutVolume( key );
-                float max {0};
-		for ( jack_nframes_t n=0; n<nframes; n++ ) {
-                        tmp[ n ] *= volume;
-                        max = qMax(max, static_cast<float>(tmp[n]));
-                }
-                backend->newOutputLevel(key, max);
+		float volume_new = backend->getOutVolumeNew( key );
+		float max {0};
+		if ( qFabs( ( volume_new - volume ) / volume ) > threshold ) {
+			qDebug() << "BIG CHANGE";
+			for ( jack_nframes_t n=0; n<nframes; n++ ) {
+				tmp[ n ] *=	volume + n * ( volume_new - volume ) / nframes;
+				max = qMax( max, static_cast<float>( tmp[ n ] ) );
```

Comment:
It's been a while since I worked on this (or in this language;-) ), but when you cast `jack_default_audio_sample_t` to `float`, this is only save where both are the same. Wouldn't it be better to define `jack_default_audio_sample_t max {0};` and only convert (not cast) to float when needed?

Reply:
I checked the C++ prototype of qMax and it's a template with the same type for both args. I'm out of date with C++ metaprogramming, but as we are dealing with native numerical types (`float`, `double`, perhaps `int` or `short`) you might not need a cast at all; type promotion should promote the lesser type automatically? So if you say `qMax(1, 2.0)`  I think the only way that template can work is to use `double` for both arguments, meaning that compiler will have to promote the integer argument to double before passing it to the `double qMax(double, double)` instance of the template.

I'll check this out before signing off on it.